### PR TITLE
Use ed25519_sign_open_batch() to verify multisignatures:

### PR DESCRIPTION
**Pull request by mrtcnk**
## High Level Overview of Change

Measurements show that batch verification of ed25519 signatures can reduce execution time by nearly 50% compared to sequential verification.  secp256k1 signatures are unaffected.

### Context of Change

Batch verification for Ed25519 signatures is added.

### Type of Change

- [ x ] New feature (non-breaking change which adds functionality)
- [ x ] Tests (You added tests for code that already exists, or your new feature included in this PR)

### CODE REVIEW

**Comment by intelliot** [comment]
note: @scottschurr contributed to this code and would not be an unbiased reviewer. require 2 other reviewers, including at least one good C++ reviewer.

File: src/ripple/protocol/impl/PublicKey.cpp [hunk 1]
```diff
@@ -299,6 +299,111 @@ verify(
...
+    }
+    // If there are no ed25519 signatures, then we're done.
+    if (verifyResultEd.empty())
+        return result;
+
+    // Batch verify the ed25519 signatures since it can save us some time.
+    std::vector<int> validSigsEd(messagesEd.size(), 0);
+    try
+    {
+        ed25519_sign_open_batch(
```
Comment (Bronek) [inline src/ripple/protocol/impl/PublicKey.cpp:383]: Our of curiosity, `mustBeFullyCanonical` does not apply to ed25519 signatures ?

Comment (scottschurr) [reply src/ripple/protocol/impl/PublicKey.cpp:383]: Great catch!  @mrtcnk and I were so focused on the signature part that I missed this.

As it turns out, `mustBeFullyCanonical` kind of doesn't apply, since rippled requires _all_ ed25519 signatures to be canonical.  But the canonical check was lost in this refactor.

A good fix might be to put the following code into the `verifyBatch()` implementation:
```
        else if (*type == KeyType::ed25519)
        {
            // rippled unconditionally requires all ed25519 signatures to be
            // canonical, and batch verification won't do that work for us.
            if (!ed25519Canonical(makeSlice(data[i].signature)))
            {
                result[i] = sigInvalid;
                return result;
            }

            // If the type is ed25519, then include it in the vectors
            // that will be sent to the batch function later.
            messagesEd.push_back(data[i].message.peekData().data());
 ...
```

**Review by Bronek — CHANGES_REQUESTED** [review]
Enforce canonical `ed25519` signatures; other inline remarks from me are non-essential

File: src/ripple/protocol/PublicKey.h [hunk 2]
```diff
@@ -259,6 +259,33 @@ verify(
...
+    Serializer message;
+    Blob signature;
+
+    explicit VerifyBatchData(PublicKey const& publicKey) : pubKey(publicKey)
+    {
+    }
+};
+
+/** Result returned from verifyBatch */
+enum class VerifyResult { sigUnexamined = 0, sigValid, sigInvalid };
```
[the remark this answers is not in the document]
Comment (scottschurr) [reply src/ripple/protocol/PublicKey.h:275]: Yeah, I'm not convinced there's much value here.  What really matters is whether the signature is valid or invalid.  For our purposes a non-canonical signature is still invalid.

**Comment by mrtcnk** [comment]
Thank you @Bronek and @scottschurr for your comments. I have addressed them in the most recent commit.

**Review by Bronek — APPROVED** [review]

File: src/ripple/protocol/impl/STTx.cpp [hunk 3]
```diff
@@ -359,63 +361,73 @@ STTx::checkMultiSign(
...
+        verifyBatch(batchVerifyData, fullyCanonical);
+
+    // The result should always equal batchVerifyData.size(), so this should
+    // never happen.
+    if (result.size() != signers.size())
+        return Unexpected("Invalid signature.");
+
+    // Return the result of the verification.
+    bool hasUnexamined = false;
+    for (std::size_t i = 0; i < signers.size(); ++i)
```
[the remark this answers is not in the document]
Comment (scottschurr) [reply src/ripple/protocol/impl/STTx.cpp:410]: Thanks for pointing it out in the first place.  If we didn't, every one of us, miss details then there would be no need for code reviews.  I appreciate your careful review.

**Review by nbougalis — CHANGES_REQUESTED** [review]
**Veto**. This change **MUST NOT** go in.

First, I think that this code is unlikely to translate to any meaningful performance benefits in real-world scenarios which makes the additional complexity introduced here difficult to justify.

But more importantly, it's broken in weird and subtle ways that aren't immediately obvious because batch Ed25519 verification is complex and  complicated.

Instead of me typing out all the issues that this change opens up, please refer to this truly excellent and comprehensive blog post by @hdevalence: https://hdevalence.ca/blog/2020-10-04-its-25519am.

**Comment by sublimator** [comment]
The link got messed up: https://hdevalence.ca/blog/2020-10-04-its-25519am

**Comment by Bronek** [comment]
@nbougalis thank you for the very informative link.

Not commenting on the complexity or expected benefits, if I read correctly the one way to ensure that the batch equation does not break the blockchain is to use batch equation for **both** single and batch signatures, whilst ensuring that the invalid set of signatures for batch equation is **strictly** a subset of invalid signatures for the currently used equation ?

This is of course made more complex by the fact that the donna implementation which we currently use will automatically select **unbatched** equation for 3 or fewer signatures - even if batch is actually larger than that, and the 3 (or fewer) signatures are leftovers after batch verification of blocks sized `max_batch_size`.

Finally, I think we might want to use similar matrix of edge case signatures in our own unit tests, to ensure that the signatures deemed **valid** as of today do not suddenly become **invalid** in the future.

**Comment by hdevalence** [comment]
I have no context on the project so this is a drive-by comment.

That said, using or not using batch verification does not absolve a protocol from having subtle consensus-related behaviors around signature verification. You have to specify an exact set of rules anyways. If you do that, you may as well specify the ZIP-215 rules described in that post, in which case there's no problem with batch verification.

**Comment by nbougalis** [comment]
> Not commenting on the complexity or expected benefits, if I read correctly the one way to ensure that the batch equation does not break the blockchain consensus is to use batch equation for **both** single and batch signatures, whilst ensuring that the invalid set of signatures for batch equation is **strictly** a subset of invalid signatures for the currently used equation ? Which BTW seem to be a property of batched equation in donna implementation.

It's a little more complex than that, unfortunately. Batched or unbatched equations, there are still some issues, which are _partially_ "mitigated" (quotes intentional) by the fact that there's only one implementation of the protocol and it uses Donna so the behavior is "consistent". But that's not a good place to be and it's worth considering whether more stringent checking of signature canonicality is warranted.

I need to think this through a little more carefully before offering an opinion, but I'd be curious to hear from @JoelKatz and, (if he's available) @hdevalence.

**Comment by JoelKatz** [comment]
Hey, @nbougalis, thanks for your comments. You are absolutely right that this is not acceptable as it is. This change must be triggered by an amendment.

Then, it will not introduce any new problems. Already, you must precisely match rippled's signature verification logic if you want to be assured the same results. That will still be the case.

If we do make this change, we will be changing the signature algorithm slightly. At the same time, it might be a good idea to tighten up our verification rules to make interoperability less painful.

The performance impact is measureable in larger multisignature cases. Maybe if we tighten up the signature verification rules, that alone can justify making a change, and the performance improvements would just be a bonus.

**Comment by scottschurr** [comment]
@mrtcnk, I have heard that this pull request is on hold until other work is completed.  Please consider closing this pull request until the other work is complete.  You can re-open this pull request once the other work is completed.  Thanks.
